subprojects/execution/src/main/java/org/gradle/internal/execution/steps/ExecuteWorkBuildOperationFiringStep.java

```diff
@@ -39,29 +40,31 @@ public ExecuteWorkBuildOperationFiringStep(BuildOperationExecutor buildOperation
 
     @Override
     public R execute(UnitOfWork work, C context) {
-        return operation(operationContext -> {
-                R result = delegate.execute(work, context);
-                ExecuteWorkBuildOperationType.Result operationResult = new ExecuteWorkResult(result.getExecution());
-                operationContext.setResult(operationResult);
-                return result;
-            },
-            BuildOperationDescriptor
-                .displayName("Execute unit of work")
-                .details(new ExecuteWorkDetails(work)));
+        Optional<String> buildOperationWorkType = work.getBuildOperationWorkType();
+        return buildOperationWorkType.map(workType -> operation(operationContext -> {
+                    R result = delegate.execute(work, context);
+                    ExecuteWorkBuildOperationType.Result operationResult = new ExecuteWorkResult(result.getExecution());
+                    operationContext.setResult(operationResult);
+                    return result;
+                },
+                BuildOperationDescriptor
+                    .displayName("Execute unit of work")
+                    .details(new ExecuteWorkDetails(workType)))
+        ).orElseGet(() -> delegate.execute(work, context));
```

Comment:
:x: The whitespace here looks off. Can you please fix it?

Reply:
This is what auto-indent in IDEA gives me, so I suppose there is nothing to fix. Or are you referring to something else than the indenting?